eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec/Invalid.hs

```diff
@@ -2,228 +2,347 @@
+              isSpender = isJust . toSpendingPurpose @era @AsIx
+              removeSpenders = Map.filterWithKey (const . not . isSpender)
+              dropSpendingRedeemers = pure . (witsTxL . rdmrsTxWitsL . redeemersL %~ removeSpenders)
+          withPostFixup (dropSpendingRedeemers >=> fixupPPHash >=> resetAddrWits) $
+            withPPRTFailures $
+              submitFailingTx
+                tx
+                [ injectFailure $
+                    ExtraRedeemers [mkMintingPurpose $ AsIx 0]
+                , injectFailure $
+                    MissingRedeemers [(mkSpendingPurpose $ AsItem txIn, scriptHash)]
+                , injectFailure $
+                    CollectErrors [NoRedeemer $ mkSpendingPurpose $ AsItem txIn]
+                ]
 
-  it "Insufficient collateral" $ do
-    const $ pendingWith "not implemented yet"
+        it "Missing witness for collateral input" $ do
+          let scriptHash = alwaysSucceedsWithDatumHash
+          scriptInput <- produceScript scriptHash
+          (collateralHash, collateralAddr) <- freshKeyAddr
+          collateralInput <- sendCoinTo collateralAddr $ Coin 1_000_000
+          let
+            tx =
+              mkBasicTx mkBasicTxBody
+                & bodyTxL . inputsTxBodyL <>~ [scriptInput]
+                & bodyTxL . collateralInputsTxBodyL <>~ [collateralInput]
+            isOtherWitness wit = witVKeyHash wit /= asWitness collateralHash
+            dropCollateralWitness = witsTxL . addrTxWitsL %~ Set.filter isOtherWitness
+          withPostFixup (pure . dropCollateralWitness) $
+            submitFailingTx
+              tx
+              [injectFailure $ MissingVKeyWitnessesUTXOW [asWitness collateralHash]]
 
-  it "Unacceptable extra redeemer" $ do
-    const $ pendingWith "not implemented yet"
+        -- Post-Alonzo eras produce additional post-Alonzo predicate failures that we can't include here
+        unless (lang > eraMaxLanguage @Alonzo) $ do
+          describe "Extra Redeemer" $ do
+            let
+              testPurpose purpose = do
+                txIn <- produceScript alwaysSucceedsWithDatumHash
+                let redeemer = (Data $ P.I 442, ExUnits 443 444) -- Needs to be unique
+                    tx =
+                      mkBasicTx mkBasicTxBody
+                        & bodyTxL . inputsTxBodyL <>~ [txIn]
+                        & witsTxL . rdmrsTxWitsL . redeemersL <>~ Map.singleton purpose redeemer
+                txFixed <- fixupTx tx
+                -- The `Ix` of the redeemer may have been changed by `fixupRedeemerIndices`
+                let fixedRedeemers = txFixed ^. witsTxL . rdmrsTxWitsL . redeemersL
+                    extraRedeemers = Map.keys $ Map.filter (== redeemer) fixedRedeemers
+                withNoFixup $
+                  withPPRTFailures $
+                    submitFailingTx
+                      txFixed
+                      [injectFailure $ ExtraRedeemers extraRedeemers]
```

Comment:
I tried this previously, and the problem is that the "other failures" are Babbage- and Conway-specific. That would mean importing Babbage and Conway modules, and I assume we don't want to do that in an Alonzo test module. I don't think it would produce circular imports because we'd be importing core modules rather than test modules, but it still seems like bad design.

An alternative would be to perform a fixup on the returned failures, and somehow filter out the non-Alonzo ones. However, I'm not sure how to do that, or whether it's even possible.

Reply:
The additional failures are all like this one:

```
UtxowFailure
  (UtxoFailure
    (AlonzoInBabbageUtxoPredFailure
      (UtxosFailure
        (CollectErrors [BadTranslation (RedeemerPointerPointsToNothing (AlonzoMinting 2))]))))]
```

`RedeemerPointerPointsToNothing` is defined in `Cardano.Ledger.Babbage.TxInfo`.

If I make the redeemer point to something, the redeemers are no longer "extra" and the transaction succeeds.

Reply:
BTW, the "Multiple equal plutus-locked certs" test fails in Conway but not in Babbage. "Minting" and "Spending" fail in both Babbage and Conway.

Reply:
An easy option, although not a satisfying one, would be to move these tests to Conway and build the failure lists conditionally there, using the types that are already in scope.

Reply:
I see. That makes total sense. 
In that case I think we just need to add a test like that to Babbage imp tests that checks this for Babbage onwards for PlutusV2 and another one that checks for Conway onwards with PlutusV3.
That being said we don't have to do that in this PR. If you can add those as par to of the later work you'll be doing that would be super helpful. Feel free to convert this comment into an issue if you'd like to keep track of it on github.

Reply:
Makes sense. Will do.